operator/controller/dnsnameresolver/resolver.go

```diff
@@ -114,8 +114,9 @@ func (resolver *Resolver) Start() {
 				// after the remaining duration.
 				timeTillNextLookup = remainingDuration
 			} else {
-				// TTL of the DNS name has already expired, so send DNS lookup request as soon as possible.
-				timeTillNextLookup = 1 * time.Millisecond
+				// A DNS lookup request has been sent upon TTL expiration of the DNS name. Reset the timer to wait until twice of default
+				// minimum TTL to perform the next lookup.
+				timeTillNextLookup = 2 * defaultMinTTL
```

Comment:
Also, it would be nice to group the logic of the next lookup computation into a dedicated function which would `select` on the channels and would give the next lookup time. This would allow us to unit test it and clearly see the use cases.

Reply:
> I'm not sure I'm getting this one. Do we kinda let the expired records to be removed instead of trying to save them resending the dns request?

When the next lookup time is reached, we get a signal on the `timer.C` channel in line no. 88 . The lookup for the DNS name happens in line no. 92. So when we reach here, the lookup request is already sent and we should wait for the update event for the corresponding DNSNameResolver object. 

When we receive the update event, we also remove the expired IP addresses after a grace period of `defaultMinTTL` during reconciliation. The following code does this work:
https://github.com/openshift/coredns-ocp-dnsnameresolver/blob/a3f67ee3f938fbd7f63ee7169019b46314e44da5/operator/controller/dnsnameresolver/controller.go#L139-L153

So, this block of code will not receive the new IPs/TTLs until the expired IPs are removed. If it doesn't wait until `2 * defaultMinTTL`, unnecessary lookup requests will again be sent to the CoreDNS pods.

Reply:
Separated the code for computing time till next lookup into a function and added unit test for the same.

Reply:
> When the next lookup time is reached, we get a signal on the timer.C channel in line no. 88 . The lookup for the DNS name happens in line no. 92. So when we reach here, the lookup request is already sent and we should wait for the update event for the corresponding DNSNameResolver object.

But we get into this condition only when the remaining duration is `0`.  The case you added confirms this:
```go
		{
			name:                       "DNS exists and remaining duration is not greater than 0",
			dnsExists:                  true,
			remainingDuration:          0,
			expectedTimeTillNextLookup: 2 * defaultMinTTL,
		},
```

So something doesn't stick here. The remaining duration says that we need to send a new lookup asap (like it was before this PR) but we delay it to catchup with the updates.